writeln_empty_string contains hardcoded suggestion

<!--
Hi there! Whether you've come to make a suggestion for a new lint, an improvement to an existing lint or to report a bug or a false positive in Clippy, you've come to the right place.

If you want to report that Clippy does not compile, please be sure to be using the *latest version* of *Rust nightly*! Compiler plugins are highly unstable and will only work with a nightly Rust for now. If you are but still have a problem, please let us now!

Thank you for using Clippy!

Write your comment below this line: -->
Actual output:
```
warning: using `writeln!(v, "")`
   --> src/main.rs:106:9
    |
106 |         writeln!(out, "").unwrap();
    |         ^^^^^^^^^^^^^^^^^ help: replace it with: `writeln!(v)`
    |
    = note: #[warn(writeln_empty_string)] on by default
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#writeln_empty_string
```
Expected output:
```
warning: using `writeln!(out, "")`
   --> src/main.rs:106:9
    |
106 |         writeln!(out, "").unwrap();
    |         ^^^^^^^^^^^^^^^^^ help: replace it with: `writeln!(out)`
    |
    = note: #[warn(writeln_empty_string)] on by default
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#writeln_empty_string
```
See lines:
https://github.com/rust-lang-nursery/rust-clippy/pull/2630/files#diff-2e8e94b55639a4808cb93bcc616d2d92R238
https://github.com/rust-lang-nursery/rust-clippy/pull/2630/files#diff-2e8e94b55639a4808cb93bcc616d2d92R240

cc @oli-obk for missing it during review :rofl:

I did get somewhere, but I am clueless how to solve the issues with `?` and why this is not working:
```
error[E0599]: no method named `into_vec` found for type `syntax::ptr::P<syntax::ast::Expr>` in the current scope
   --> clippy_lints/src/write.rs:240:27
    |
240 |         Some(p) => Some(p.into_vec().0),
    |                           ^^^^^^^^

error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
   --> clippy_lints/src/write.rs:244:5
    |
244 |     parser.expect(&token::Comma).map_err(|mut err| err.cancel()).ok()?;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `(std::option::Option<std::string::String>, std::option::Option<syntax::ast::Expr>)`
```
The documentation for `syntax::ptr::P` says that there should be a `into_vec` and I think that's the only way to extract the `Expr` from this smart pointer.
Also should I remove the if for `is_write` and leave the `parse_expr` where it is?

--- a/clippy_lints/src/write.rs
+++ b/clippy_lints/src/write.rs
@@ -219,7 +219,7 @@ impl EarlyLintPass for Pass {
                         cx,
                         WRITELN_EMPTY_STRING,
                         mac.span,
-                        format!("using writeln!({}, \"\")", suggestion).as_str(),
+                        format!("using `writeln!({}, \"\")`", suggestion).as_str(),
                         "replace it with",
                         format!("writeln!({})", suggestion),
                     );

--- a/clippy_lints/src/write.rs
+++ b/clippy_lints/src/write.rs
@@ -3,6 +3,7 @@ use rustc::{declare_lint, lint_array};
 use syntax::ast::*;
 use syntax::tokenstream::{ThinTokenStream, TokenStream};
 use syntax::parse::{token, parser};
+use std::borrow::Cow;
 use crate::utils::{span_lint, span_lint_and_sugg, snippet};
 
 /// **What it does:** This lint warns when you use `println!("")` to
@@ -212,7 +213,7 @@ impl EarlyLintPass for Pass {
             let check_tts = check_tts(cx, &mac.node.tts, true);
             if let Some(fmtstr) = check_tts.0 {
                 if fmtstr == "" {
-                    let suggestion = check_tts.1.map_or("v", |expr| snippet(cx, expr.span, "v").into_owned().as_str());
+                    let suggestion = check_tts.1.map_or(Cow::Borrowed("v"), |expr| snippet(cx, expr.span, "v"));
 
                     span_lint_and_sugg(
                         cx,
@@ -220,7 +221,7 @@ impl EarlyLintPass for Pass {
                         mac.span,
                         format!("using writeln!({}, \"\")", suggestion).as_str(),
                         "replace it with",
-                        format!("writeln!({})", "v"),
+                        format!("writeln!({})", suggestion),
                     );
                 }
             }
@@ -239,20 +240,19 @@ fn check_tts<'a>(cx: &EarlyContext<'a>, tts: &ThinTokenStream, is_write: bool) -
     );
     let mut expr: Option<Expr> = None;
     if is_write {
-        // skip the initial write target
-        expr = match parser.parse_expr().map_err(|mut err| err.cancel()).ok() {
-            Some(p) => Some(p.and_then(|expr| expr)),
-            None => return (None, None),
+        expr = match parser.parse_expr().map_err(|mut err| err.cancel()) {
+            Ok(p) => Some(p.into_inner()),
+            Err(_) => return (None, None),
         };
         // might be `writeln!(foo)`
-        if let None = parser.expect(&token::Comma).map_err(|mut err| err.cancel()).ok() {
+        if parser.expect(&token::Comma).map_err(|mut err| err.cancel()).is_err() {
             return (None, expr);
         }
     }
 
-    let fmtstr = match parser.parse_str().map_err(|mut err| err.cancel()).ok() {
-        Some(token) => token.0.to_string(),
-        None => return (None, expr),
+    let fmtstr = match parser.parse_str().map_err(|mut err| err.cancel()) {
+        Ok(token) => token.0.to_string(),
+        Err(_) => return (None, expr),
     };
     use fmt_macros::*;
     let tmp = fmtstr.clone();
@@ -281,9 +281,9 @@ fn check_tts<'a>(cx: &EarlyContext<'a>, tts: &ThinTokenStream, is_write: bool) -
             assert!(parser.eat(&token::Eof));
             return (Some(fmtstr), expr);
         }
-        let token_expr = match parser.parse_expr().map_err(|mut err| err.cancel()).ok() {
-            Some(expr) => expr,
-            None => return (Some(fmtstr), None),
+        let token_expr = match parser.parse_expr().map_err(|mut err| err.cancel()) {
+            Ok(expr) => expr,
+            Err(_) => return (Some(fmtstr), None),
         };
         const SIMPLE: FormatSpec<'_> = FormatSpec {
             fill: None,

--- a/clippy_lints/src/write.rs
+++ b/clippy_lints/src/write.rs
@@ -3,7 +3,7 @@ use rustc::{declare_lint, lint_array};
 use syntax::ast::*;
 use syntax::tokenstream::{ThinTokenStream, TokenStream};
 use syntax::parse::{token, parser};
-use crate::utils::{span_lint, span_lint_and_sugg};
+use crate::utils::{span_lint, span_lint_and_sugg, snippet};
 
 /// **What it does:** This lint warns when you use `println!("")` to
 /// print a newline.
@@ -212,13 +212,15 @@ impl EarlyLintPass for Pass {
             let check_tts = check_tts(cx, &mac.node.tts, true);
             if let Some(fmtstr) = check_tts.0 {
                 if fmtstr == "" {
+                    let suggestion = check_tts.1.map_or("v", |expr| snippet(cx, expr.span, "v").into_owned().as_str());
+
                     span_lint_and_sugg(
                         cx,
                         WRITELN_EMPTY_STRING,
                         mac.span,
-                        format!("using `writeln!({}, \"\")`", check_tts.1).as_str(),
+                        format!("using writeln!({}, \"\")", suggestion).as_str(),
                         "replace it with",
-                        format!("using `writeln!({})`", check_tts.1),
+                        format!("writeln!({})", "v"),
                     );
                 }
             }
@@ -235,15 +237,23 @@ fn check_tts<'a>(cx: &EarlyContext<'a>, tts: &ThinTokenStream, is_write: bool) -
         false,
         false,
     );
-    // skip the initial write target
-    let expr: Option<Expr> = match parser.parse_expr().map_err(|mut err| err.cancel()).ok() {
-        Some(p) => Some(p.into_vec().0),
-        None => None,
-    };
-    // might be `writeln!(foo)`
-    parser.expect(&token::Comma).map_err(|mut err| err.cancel()).ok()?;
+    let mut expr: Option<Expr> = None;
+    if is_write {
+        // skip the initial write target
+        expr = match parser.parse_expr().map_err(|mut err| err.cancel()).ok() {
+            Some(p) => Some(p.and_then(|expr| expr)),
+            None => return (None, None),
+        };
+        // might be `writeln!(foo)`
+        if let None = parser.expect(&token::Comma).map_err(|mut err| err.cancel()).ok() {
+            return (None, expr);
+        }
+    }
 
-    let fmtstr = parser.parse_str().map_err(|mut err| err.cancel()).ok()?.0.to_string();
+    let fmtstr = match parser.parse_str().map_err(|mut err| err.cancel()).ok() {
+        Some(token) => token.0.to_string(),
+        None => return (None, expr),
+    };
     use fmt_macros::*;
     let tmp = fmtstr.clone();
     let mut args = vec![];
@@ -271,7 +281,10 @@ fn check_tts<'a>(cx: &EarlyContext<'a>, tts: &ThinTokenStream, is_write: bool) -
             assert!(parser.eat(&token::Eof));
             return (Some(fmtstr), expr);
         }
-        let expr = parser.parse_expr().map_err(|mut err| err.cancel()).ok()?;
+        let token_expr = match parser.parse_expr().map_err(|mut err| err.cancel()).ok() {
+            Some(expr) => expr,
+            None => return (Some(fmtstr), None),
+        };
         const SIMPLE: FormatSpec<'_> = FormatSpec {
             fill: None,
             align: AlignUnknown,
@@ -280,7 +293,7 @@ fn check_tts<'a>(cx: &EarlyContext<'a>, tts: &ThinTokenStream, is_write: bool) -
             width: CountImplied,
             ty: "",
         };
-        match &expr.node {
+        match &token_expr.node {
             ExprKind::Lit(_) => {
                 let mut all_simple = true;
                 let mut seen = false;
@@ -296,7 +309,7 @@ fn check_tts<'a>(cx: &EarlyContext<'a>, tts: &ThinTokenStream, is_write: bool) -
                     }
                 }
                 if all_simple && seen {
-                    span_lint(cx, lint, expr.span, "literal with an empty format string");
+                    span_lint(cx, lint, token_expr.span, "literal with an empty format string");
                 }
                 idx += 1;
             },

--- a/clippy_lints/src/write.rs
+++ b/clippy_lints/src/write.rs
@@ -179,7 +179,7 @@ impl EarlyLintPass for Pass {
     fn check_mac(&mut self, cx: &EarlyContext<'_>, mac: &Mac) {
         if mac.node.path == "println" {
             span_lint(cx, PRINT_STDOUT, mac.span, "use of `println!`");
-            if let Some(fmtstr) = check_tts(cx, &mac.node.tts, false) {
+            if let Some(fmtstr) = check_tts(cx, &mac.node.tts, false).0 {
                 if fmtstr == "" {
                     span_lint_and_sugg(
                         cx,
@@ -193,7 +193,7 @@ impl EarlyLintPass for Pass {
             }
         } else if mac.node.path == "print" {
             span_lint(cx, PRINT_STDOUT, mac.span, "use of `print!`");
-            if let Some(fmtstr) = check_tts(cx, &mac.node.tts, false) {
+            if let Some(fmtstr) = check_tts(cx, &mac.node.tts, false).0 {
                 if fmtstr.ends_with("\\n") && !fmtstr.ends_with("\\n\\n") {
                     span_lint(cx, PRINT_WITH_NEWLINE, mac.span,
                             "using `print!()` with a format string that ends in a \
@@ -201,7 +201,7 @@ impl EarlyLintPass for Pass {
                 }
             }
         } else if mac.node.path == "write" {
-            if let Some(fmtstr) = check_tts(cx, &mac.node.tts, true) {
+            if let Some(fmtstr) = check_tts(cx, &mac.node.tts, true).0 {
                 if fmtstr.ends_with("\\n") && !fmtstr.ends_with("\\n\\n") {
                     span_lint(cx, WRITE_WITH_NEWLINE, mac.span,
                             "using `write!()` with a format string that ends in a \
@@ -209,15 +209,16 @@ impl EarlyLintPass for Pass {
                 }
             }
         } else if mac.node.path == "writeln" {
-            if let Some(fmtstr) = check_tts(cx, &mac.node.tts, true) {
+            let check_tts = check_tts(cx, &mac.node.tts, true);
+            if let Some(fmtstr) = check_tts.0 {
                 if fmtstr == "" {
                     span_lint_and_sugg(
                         cx,
                         WRITELN_EMPTY_STRING,
                         mac.span,
-                        "using `writeln!(v, \"\")`",
+                        format!("using `writeln!({}, \"\")`", check_tts.1).as_str(),
                         "replace it with",
-                        "writeln!(v)".to_string(),
+                        format!("using `writeln!({})`", check_tts.1),
                     );
                 }
             }
@@ -225,7 +226,7 @@ impl EarlyLintPass for Pass {
     }
 }
 
-fn check_tts<'a>(cx: &EarlyContext<'a>, tts: &ThinTokenStream, is_write: bool) -> Option<String> {
+fn check_tts<'a>(cx: &EarlyContext<'a>, tts: &ThinTokenStream, is_write: bool) -> (Option<String>, Option<Expr>) {
     let tts = TokenStream::from(tts.clone());
     let mut parser = parser::Parser::new(
         &cx.sess.parse_sess,
@@ -234,12 +235,14 @@ fn check_tts<'a>(cx: &EarlyContext<'a>, tts: &ThinTokenStream, is_write: bool) -
         false,
         false,
     );
-    if is_write {
-        // skip the initial write target
-        parser.parse_expr().map_err(|mut err| err.cancel()).ok()?;
-        // might be `writeln!(foo)`
-        parser.expect(&token::Comma).map_err(|mut err| err.cancel()).ok()?;
-    }
+    // skip the initial write target
+    let expr: Option<Expr> = match parser.parse_expr().map_err(|mut err| err.cancel()).ok() {
+        Some(p) => Some(p.into_vec().0),
+        None => None,
+    };
+    // might be `writeln!(foo)`
+    parser.expect(&token::Comma).map_err(|mut err| err.cancel()).ok()?;
+
     let fmtstr = parser.parse_str().map_err(|mut err| err.cancel()).ok()?.0.to_string();
     use fmt_macros::*;
     let tmp = fmtstr.clone();
@@ -247,7 +250,7 @@ fn check_tts<'a>(cx: &EarlyContext<'a>, tts: &ThinTokenStream, is_write: bool) -
     let mut fmt_parser = Parser::new(&tmp, None);
     while let Some(piece) = fmt_parser.next() {
         if !fmt_parser.errors.is_empty() {
-            return None;
+            return (None, expr);
         }
         if let Piece::NextArgument(arg) = piece {
             if arg.format.ty == "?" {
@@ -266,7 +269,7 @@ fn check_tts<'a>(cx: &EarlyContext<'a>, tts: &ThinTokenStream, is_write: bool) -
     loop {
         if !parser.eat(&token::Comma) {
             assert!(parser.eat(&token::Eof));
-            return Some(fmtstr);
+            return (Some(fmtstr), expr);
         }
         let expr = parser.parse_expr().map_err(|mut err| err.cancel()).ok()?;
         const SIMPLE: FormatSpec<'_> = FormatSpec {

--- a/tests/ui/writeln_empty_string.rs
+++ b/tests/ui/writeln_empty_string.rs
@@ -5,8 +5,11 @@ use std::io::Write;
 fn main() {
     let mut v = Vec::new();
 
-    // This should fail
+    // These should fail
     writeln!(&mut v, "");
+
+    let mut suggestion = Vec::new();
+    writeln!(&mut suggestion, "");
 
     // These should be fine
     writeln!(&mut v);

--- a/tests/ui/writeln_empty_string.stderr
+++ b/tests/ui/writeln_empty_string.stderr
@@ -1,10 +1,16 @@
-error: using `writeln!(v, "")`
+error: using writeln!(&mut v, "")
  --> $DIR/writeln_empty_string.rs:9:5
   |
 9 |     writeln!(&mut v, "");
-  |     ^^^^^^^^^^^^^^^^^^^^ help: replace it with: `writeln!(v)`
+  |     ^^^^^^^^^^^^^^^^^^^^ help: replace it with: `writeln!(&mut v)`
   |
   = note: `-D writeln-empty-string` implied by `-D warnings`
 
-error: aborting due to previous error
+error: using writeln!(&mut suggestion, "")
+  --> $DIR/writeln_empty_string.rs:12:5
+   |
+12 |     writeln!(&mut suggestion, "");
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `writeln!(&mut suggestion)`
+
+error: aborting due to 2 previous errors
 

--- a/tests/ui/writeln_empty_string.stderr
+++ b/tests/ui/writeln_empty_string.stderr
@@ -1,4 +1,4 @@
-error: using writeln!(&mut v, "")
+error: using `writeln!(&mut v, "")`
  --> $DIR/writeln_empty_string.rs:9:5
   |
 9 |     writeln!(&mut v, "");
@@ -6,7 +6,7 @@ error: using writeln!(&mut v, "")
   |
   = note: `-D writeln-empty-string` implied by `-D warnings`
 
-error: using writeln!(&mut suggestion, "")
+error: using `writeln!(&mut suggestion, "")`
   --> $DIR/writeln_empty_string.rs:12:5
    |
 12 |     writeln!(&mut suggestion, "");

